### PR TITLE
Efficient implementation of `Tensor::ones()` for `metal`

### PR DESCRIPTION
**Pull request by AnubhabB**
While working with `Tensor::ones()` I realized the implementation is significantly slower than `pytorch` counterpart in `metal`. The original implementation would create a `CpuStorage` and move it to Gpu.

This implementation introduces a metal kernel function for `Tensor::ones()` impl.

### Benchmarks (original vs new implementation)
[Benchmark Code](https://github.com/AnubhabB/tensor-bench/blob/ones-bench/benches/benchmarks/ones.rs) ran on Mac M1 Pro 16GB

| Impl            | DType  | Low        | Mid        | High       | Throughput (GiB/s) |
|:----------------:|:------:|:----------:|:----------:|:----------:|:---------------:|
| Original(cpu)       | `u8`   | 36.752 µs  | 39.081 µs  | 40.932 µs  | 24.988             |
| Original(metal)    | `u8`   | 226.73 µs  | 230.18 µs  | 233.76 µs  | 4.2427             |
| New Impl(metal) | `u8`   | 14.773 µs  | 14.877 µs  | 14.970 µs  | 4.2427             |
| Original(cpu)   | `u32`  | 349.63 µs  | 352.22 µs  | 355.47 µs  | 11.090             |
| Original(metal) | `u32`  | 1.2362 ms  | 1.2555 ms  | 1.2761 µs  | 3.1112             |
| New Impl(metal) | `u32`  | 25.443 µs  | 25.515 µs  | 25.572 µs  | 153.09             |
| Original(cpu)   | `f32`  | 351.63 µs  | 355.10 µs  | 359.24 µs  | 11.001             |
| Original(metal) | `f32`  | 1.2358 ms  | 1.2552 ms  | 1.2800 ms  | 3.1121             |
| New Impl(metal) | `f32`  | 24.859 µs  | 24.955 µs  | 25.039 µs  | 156.53             |
| Original(cpu)   | `bf16` | 179.12 µs  | 181.66 µs  | 184.41 µs  | 10.752             |
| Original(metal) | `bf16` | 634.81 µs  | 650.41 µs  | 667.20 µs  | 3.0029             |
| New Impl(metal) | `bf16` | 16.683 µs  | 16.782 µs  | 16.871 µs  | 116.38             |
| Original(cpu)   | `f16`  | 175.30 µs  | 177.45 µs  | 180.00 µs  | 11.007             |
| Original(metal) | `f16`  | 636.60 µs  | 648.13 µs  | 662.07 µs  | 3.0135             |
| New Impl(metal) | `f16`  | 16.710 µs  | 16.767 µs  | 16.818 µs  | 116.49             |
| Original(cpu)   | `i64`  | 703.17 µs  | 708.68 µs  | 714.85 µs  | 11.024             |
| Original(metal) | `i64`  | 2.2823 ms  | 2.3245 ms  | 2.3691 ms  | 3.3609             |
| New Impl(metal) | `i64`  | 42.224 µs  | 42.342 µs  | 42.437 µs  | 184.51             |

Looks like we are getting some significant performance boost.

### Implementation Details:

- `MetalDevice::ones_impl` now introduces a code-path similar to the `random` implementation - falls back to `CpuStorage` based implementation for `DType::f64`
- Introduces a kernel `fill` - following the pattern implemented for `normal` metal kernels
- Expanded `candle-core` test cases to test for `bf16` and `f16` types
- Added test-case to `candle-metal-kernels` for the `constant_fill` related code introduced

Please let me know if I'm missing something!

### CODE REVIEW

File: candle-metal-kernels/src/lib.rs [hunk 1]
```diff
@@ -2357,5 +2360,32 @@ pub fn call_mlx_gemm(
...
+    let encoder = ep.encoder();
+    let encoder: &ComputeCommandEncoderRef = encoder.as_ref();
+
+    encoder.set_compute_pipeline_state(&pipeline);
+
+    set_params!(encoder, (output, v));
+
+    encoder.use_resource(output, metal::MTLResourceUsage::Write);
+
+    let grid_size = MTLSize { width: length as u64, height: 1, depth: 1 };
```
[the remark this answers is not in the document]
Comment (AnubhabB) [reply candle-metal-kernels/src/lib.rs]: this was stupid of me. Resolved.

**Review by LaurentMazare — APPROVED** [review]

**Comment by AnubhabB** [comment]
@LaurentMazare `rustfmt` is faicling .. should I run a `cargo fmt --all` and push changes?

**Comment by LaurentMazare** [comment]
Yes please.

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Thanks!

**Pull request merged**